### PR TITLE
[#35] Support querying Milestone by number and add helper query to get milestone by number

### LINKED ISSUE #35: Support querying Milestone by number and add helper query to get milestone by number
Currently we don't have a way to query `Milestone` by number:

```graphql
query { 
  repository(owner: "kowainik", name: "hit-on") {
    milestone(number: 1) {
      id
    }
  }
}
```

With the response:

```json
{
  "data": {
    "repository": {
      "milestone": {
        "id": "MDk6TWlsZXN0b25lNDQ4MDI1Mg=="
      }
    }
  }
}
```

For this, we need to add a milestone connection to `Repository`. Similar to `Issue`/`Issues` we need to have `Milestone`/`Milestones`:

* https://github.com/kowainik/github-graphql/blob/19abc17c6fc1e03aa323fd22ea06164f79079452/src/GitHub/Repository.hs#L120-L126

Additionally, it would be nice to have a helper function to query `MilestoneId` similar to querying `RepositoryId`:

* https://github.com/kowainik/github-graphql/blob/19abc17c6fc1e03aa323fd22ea06164f79079452/src/GitHub/Query.hs#L84-L115

And add tests for this function (e.g. using the example in this issue).

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #35

### CODE REVIEW

File: src/GitHub/Json.hs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+However, in real-life when using queries of this library, you don't
+need to specify @"data"@ as it handled internally.
+
+The code looks like this:
+
+@
+unNested @'[ "repository", "milestone" ] @MilestoneId <$> callGitHubRaw
+@
+-}
+newtype Nested (keys :: [Symbol]) a = Nested
```
Comment (chshersh) [inline src/GitHub/Json.hs:66]: This module contains a little bit of type-level magic, but I don't know a better way at the moment 🤔 

This library is becoming more and more complex, I'm sorry for that... It's so awkward to work with GraphQL in Haskell type-safely

File: src/GitHub/Id.hs [hunk 2]
```diff
@@ -35,17 +40,21 @@ newtype Id (idType :: IdType) = Id
...
+instance Typeable idType => FromJSON (Id idType) where
+    parseJSON = withObject ("Id" <> typeName @idType) $ \o -> Id <$> (o .: "id")
 
 data IdType
     = IDAny
-    | IDRepository
+    | IDIssue
     | IDMilestone
+    | IDRepository
+    | IDUser
```
Comment (vrom911) [inline src/GitHub/Id.hs]: Let's make constructors with the lower `d` in `ID` 🙂

File: src/GitHub/Milestone.hs [hunk 3]
```diff
@@ -10,29 +10,84 @@ Data types and helper functions to work with @milestones@.
...
+milestoneToAst Milestone{..} = QueryNode
+    { queryNodeName = NodeMilestone
+    , queryNodeArgs = milestoneArgsToAst milestoneArgs
+    , queryNode     = mkQuery milestoneFieldToAst milestoneConnections
+    }
+
+{- | Arguments for the 'Milestone' connection.
+-}
+newtype MilestoneArgs (fields :: [RequiredField]) = MilestoneArgs
+    { milestoneArgsNumber    :: Int
```
Comment (vrom911) [inline src/GitHub/Milestone.hs]: Alignment here could be returned to remove spaces 🙂

File: src/GitHub/Milestone.hs [hunk 4]
```diff
@@ -53,14 +108,19 @@ milestonesToAst Milestones{..} = QueryNode
 {- | Arguments for the 'Milestones' connection.
 -}
 data MilestonesArgs (fields :: [RequiredField]) = MilestonesArgs
-    { milestonesArgsLast    :: !Int
-    , milestonesArgsOrderBy :: !(Maybe (Order MilestoneOrderField '[]))
+    { milestonesArgsLast    :: Int
```
Comment (vrom911) [inline src/GitHub/Milestone.hs:111]: We need only number of the last milestone to operate with it? Or we need an ID?

File: src/GitHub/Milestone.hs [hunk 5]
```diff
@@ -53,14 +108,19 @@ milestonesToAst Milestones{..} = QueryNode
...
+    , milestonesArgsStates  :: NonEmpty IssueState
+    , milestonesArgsOrderBy :: Maybe (Order MilestoneOrderField '[])
```
Comment (vrom911) [inline src/GitHub/Milestone.hs:113]: Interesting that we need to store this! I wonder how do we use it?

File: src/GitHub/Query.hs [hunk 6]
```diff
@@ -159,33 +197,45 @@ callGitHubRaw (GitHubToken token) text = do
...
-        pure $ GitHubDataResponse val
-
--- copy-pasted from @relude@
-typeName :: forall a. Typeable a => String
-typeName = show (typeRep @a)
+unNest
+    :: forall keys a
+    .  IO (Either GitHubError (Nested keys a))
+    -> IO (Either GitHubError a)
+unNest = fmap (second unNested)
```
Comment (vrom911) [inline src/GitHub/Query.hs:241]: Maybe Errors could be in a separate module? Or you think that its interface won't change a lot?

**Review by vrom911 — APPROVED** [review]
Outstanding job 💖

File: src/GitHub/Query.hs (hunk 6 above)
Comment (chshersh) [reply src/GitHub/Query.hs:241]: That's a great idea!

**Review by vrom911 — APPROVED** [review]

**Pull request merged**